storage/postgres/client.go

```diff
@@ -96,89 +95,99 @@ func (c *Client) SendBatch(ctx context.Context, batch *storage.QueryBatch) error
-	// However, it reports errors poorly: If _any_ query is syntactically
-	// malformed, called with the wrong number of args, or has a type conversion problem,
-	// pgx will report the _first_ query as failing.
-	//
-	// TODO: Remove the first branch if we verify that the performance gain is negligible.
-	if os.Getenv("PGX_FAST_BATCH") == "1" { //nolint:nestif
-		pgxBatch := batch.AsPgxBatch()
-		var batchResults pgx.BatchResults
-		var emptyTxOptions pgx.TxOptions
-		var tx pgx.Tx
-		var err error
-
-		// Begin a transaction.
-		useExplicitTx := opts != emptyTxOptions
-		if useExplicitTx {
-			// set up our own tx with the specified options
-			tx, err = c.pool.BeginTx(ctx, opts)
-			if err != nil {
-				return fmt.Errorf("failed to begin tx: %w", err)
-			}
-			batchResults = c.pool.SendBatch(ctx, &pgxBatch)
-		} else {
-			// use implicit tx provided by SendBatch; see https://github.com/jackc/pgx/issues/879
-			batchResults = c.pool.SendBatch(ctx, &pgxBatch)
+// Submits a new batch. Under the hood, uses `tx.SendBatch(batch.AsPgxBatch())`,
+// which is more efficient as it happens in a single roundtrip to the server.
+// However, it reports errors poorly: If _any_ query is syntactically
+// malformed, called with the wrong number of args, or has a type conversion problem,
+// pgx will report the _first_ query as failing.
+func (c *Client) sendBatchWithOptionsFast(ctx context.Context, batch *storage.QueryBatch, opts pgx.TxOptions) error {
+	pgxBatch := batch.AsPgxBatch()
+	var batchResults pgx.BatchResults
+	var emptyTxOptions pgx.TxOptions
+	var tx pgx.Tx
+	var err error
+
+	// Begin a transaction.
+	useExplicitTx := opts != emptyTxOptions
+	if useExplicitTx {
+		// set up our own tx with the specified options
+		tx, err = c.pool.BeginTx(ctx, opts)
+		if err != nil {
+			return fmt.Errorf("failed to begin tx: %w", err)
 		}
-		defer common.CloseOrLog(batchResults, c.logger)
-
-		// Exec indiviual queries in the batch.
-		for i := 0; i < pgxBatch.Len(); i++ {
-			if _, err := batchResults.Exec(); err != nil {
-				rollbackErr := ""
-				if useExplicitTx {
-					err2 := tx.Rollback(ctx)
-					if err2 != nil {
-						rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err2.Error())
-					}
+		batchResults = c.pool.SendBatch(ctx, &pgxBatch)
+	} else {
+		// use implicit tx provided by SendBatch; see https://github.com/jackc/pgx/issues/879
+		batchResults = c.pool.SendBatch(ctx, &pgxBatch)
+	}
+	defer common.CloseOrLog(batchResults, c.logger)
+
+	// Exec indiviual queries in the batch.
+	for i := 0; i < pgxBatch.Len(); i++ {
+		if _, err := batchResults.Exec(); err != nil {
+			rollbackErr := ""
+			if useExplicitTx {
+				err2 := tx.Rollback(ctx)
+				if err2 != nil {
+					rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err2.Error())
 				}
-				return fmt.Errorf("query %d %v: %w%s", i, batch.Queries()[i], err, rollbackErr)
 			}
+			return fmt.Errorf("query %d %v: %w%s", i, batch.Queries()[i], err, rollbackErr)
```

Comment:
actually is this even worse than the diagnostics we get from one-query-at-a-time mode? it's all the same info

Reply:
IIRC it always reports failure on query 0, even when the actual error appears lower down. It's been a while since I tried this though. Let me see again, and update the comment.

:exclamation: TODO

Reply:
I think [your new research](https://github.com/oasisprotocol/nexus/pull/556#issuecomment-1928184551) agrees with what I wrote. PREPARE also fails if the privileges are off (e.g. on a table), or there are issues with types. These are only expected to happen during development or experimental deploys, but it's very helpful to have a good error message in those rare cases.

Also, I just found out that when we send the results of genesis analysis to the db (= a huge tx), the fast path fails with `unexpected pipeline result: *pgconn.PipelineSync` but the slow path succeeds. I think that's a bug in pgx; postgres docs do mention that "writing clients using the pipeline mode is more complex". But I don't even want to follow up on it for such a corner case for which we even have a workaround already, inadvertently.